docs/src/main/asciidoc/_developer-guide/ejb3/EJB_on_Kubernetes.adoc

```diff
@@ -0,0 +1,105 @@
+[[EJB_on_Kubernetes]]
+= EJB on Kubernetes
+
+If the WildFly server is deployed on the Kubernetes then there are several
+points that you need to bear in mind when you use EJBs.
+
+NOTE: When deploying on Kubernetes you should consider the use of the link:https://github.com/wildfly/wildfly-operator[WildFly Operator].
+      It manages the Kubernetes objects in WildFly friendly way.
+      For example, it uses StatefulSet for the correct handling of EJB remoting and transaction recovery processing.
+
+The rest of this chapter assumes the link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/[`StatefulSet`] is used
+as the Kubernetes API object for managing the WildFly server.
+
+The StatefulSet provides a guarantee of persistent storage and network hostname stability
+over the restarts of the pod.
+
+These two guarantees are particularly important for the transaction manager which is a stateful component.
+The persistent storage over restarts is needed as the transaction log is usually stored at the file system.
+If the transaction manager creates a transaction log record it's created only at the transaction log particular to the WildFly instance.
+The hostname stability is needed as the WildFly may be contacted via EJB remote call with transaction propagation.
+The WildFly has to be reachable under the same hostname even after pod restarts.
+As the transaction log is bound to the particular WildFly instance it may be finished only there.
+
+
+[[ejb-calls-on-kubernetes]]
+== EJB calls on Kubernetes
+
+The EJB caller has two options on how to configure the remote call.
+It can be defined either as a remote outbound connection (see details at link:Admin_guide{outfilesuffix}#outbound-connections[Admin Guide, section Outbound Connections])
+or you may use a direct `InitialContext` lookup in your code.
+
+If you use either case then for the Kubernetes you need to adjust the configuration of the target node.
+For the target hostname, you need to use the DNS name of the very first pod managed by `StatefulSet`.
+
+The `StatefulSet` guarantees depend on the ordering of the pods. The pods are named in the prescribed order.
+If you scale your application up to 3 replicas you may expect to name of pods like `wildfly-server-0`, `wildfly-server-1`, `wildfly-server-2`.
+
+It's expected a link:https://kubernetes.io/docs/concepts/services-networking/service/#headless-services[headless services]
+to be used along with the `StatefulSet`. With the headless service, there is ensured the DNS hostname for the pod.
+If the WildFly is run on `StatefulSet` with the headless service named `wildfly-server-headless`.
```

Comment:
"If the application uses the WildFly Operator, a headless service will be created with a name such as `wildfly-server-headless`.